src/posix/platform/udp.cpp

```diff
@@ -0,0 +1,419 @@
+        }
+
+        if (pktinfo->ipi6_ifindex || memcmp(&pktinfo->ipi6_addr, &in6addr_any, sizeof(pktinfo->ipi6_addr)))
+        {
+            controlLength += CMSG_SPACE(sizeof(*pktinfo));
+            cmsg = CMSG_NXTHDR(&msg, cmsg);
+        }
+    }
+
+    msg.msg_controllen = controlLength;
+
+    rval = sendmsg(aFd, &msg, 0);
+    VerifyOrExit(rval > 0, perror("sendmsg"));
+
+exit:
+    return rval > 0 ? OT_ERROR_NONE : OT_ERROR_FAILED;
+}
+
+static otError receivePacket(int aFd, uint8_t *aPayload, uint16_t &aLength, otMessageInfo &aMessageInfo)
+{
+    struct sockaddr_in6 peerAddr;
+    uint8_t             control[kMaxUdpSize];
+    struct iovec        iov;
+    struct msghdr       msg;
+    ssize_t             rval;
+
+    iov.iov_base = aPayload;
+    iov.iov_len  = aLength;
+
+    msg.msg_name       = &peerAddr;
+    msg.msg_namelen    = sizeof(peerAddr);
+    msg.msg_control    = control;
+    msg.msg_controllen = sizeof(control);
+    msg.msg_iov        = &iov;
+    msg.msg_iovlen     = 1;
+    msg.msg_flags      = 0;
+
+    rval = recvmsg(aFd, &msg, 0);
+    VerifyOrExit(rval > 0, perror("recvmsg"));
+    aLength = static_cast<uint16_t>(rval);
+
+    for (struct cmsghdr *cmsg = CMSG_FIRSTHDR(&msg); cmsg != NULL; cmsg = CMSG_NXTHDR(&msg, cmsg))
+    {
+        if (cmsg->cmsg_level == IPPROTO_IPV6)
+        {
+            if (cmsg->cmsg_type == IPV6_HOPLIMIT)
+            {
+                int hoplimit           = *reinterpret_cast<int *>(CMSG_DATA(cmsg));
+                aMessageInfo.mHopLimit = hoplimit;
+            }
+            else if (cmsg->cmsg_type == IPV6_PKTINFO)
+            {
+                struct in6_pktinfo *pktinfo;
+
+                pktinfo = reinterpret_cast<in6_pktinfo *>(CMSG_DATA(cmsg));
+                memcpy(&aMessageInfo.mSockAddr, &pktinfo->ipi6_addr, sizeof(aMessageInfo.mSockAddr));
+            }
+        }
+    }
+
+    aMessageInfo.mPeerPort = ntohs(peerAddr.sin6_port);
+    memcpy(&aMessageInfo.mPeerAddr, &peerAddr.sin6_addr, sizeof(aMessageInfo.mPeerAddr));
+
+exit:
+    return rval > 0 ? OT_ERROR_NONE : OT_ERROR_FAILED;
+}
+
+otError otPlatUdpSocket(otUdpSocket *aUdpSocket)
+{
+    aUdpSocket->mHandle = kInvalidHandle;
+    return OT_ERROR_NONE;
+}
+
+otError otPlatUdpClose(otUdpSocket *aUdpSocket)
+{
+    otError error = OT_ERROR_NONE;
+    int     fd    = GetFdFromHandle(aUdpSocket->mHandle);
+
+    VerifyOrExit(fd != -1 && aUdpSocket->mSockName.mPort != 0);
+    closeFd(fd);
+
+exit:
+    aUdpSocket->mHandle = kInvalidHandle;
+    return error;
+}
+
+otError otPlatUdpBind(otUdpSocket *aUdpSocket)
+{
+    otError error = OT_ERROR_NONE;
+    int     fd    = GetFdFromHandle(aUdpSocket->mHandle);
+
+    assert(sPlatNetifIndex != 0);
+    assert(fd == -1);
+    VerifyOrExit(sPlatNetifIndex != 0, error = OT_ERROR_INVALID_STATE);
+    VerifyOrExit(fd == -1, error = OT_ERROR_ALREADY);
+
+    fd = socket(AF_INET6, SOCK_DGRAM, IPPROTO_UDP);
```

Comment:
Seems strange to create the socket within `otPlatUdpBind()`, but I understand why you did it here rather than in `otPlatUdpSocket()`. It does mean that calling `otPlatUdpBind()` is required.

Reply:
Can you clarify? OT and POSIX sockets will use an ephemeral port if `bind()` is not called before sending a message. What is the expected behavior here if `UdpSocket::SendTo()` is called without calling `UdpSocket::Bind()`?

In the case of ephemeral ports, does the platform UDP implementation allow the platform to assign the ephemeral port? If not, we need to be careful about selecting ports that may already be bound.

Reply:
Yes, there is risk that the port is already bound on either platform of OpenThread UDP. I will enhance this.